chainer/dataset/convert.py

```diff
@@ -139,3 +145,150 @@ def _concat_arrays_with_padding(arrays, padding):
+        self.stream = stream
+        self.device = None
+        self._conveyor = {}
+
+    def __call__(self, batch, device=None, padding=None):
+        """Concatenate data and transfer them to GPU asynchronously.
+
+        (see concat_examples for detail about concatenation)
+
+        Args:
+            batch (list): A list of examples.
+            device (int): Device ID to which each array is sent.
+            padding: Scalar value for extra elements.
+
+        Returns:
+            Array, a tuple of arrays, or a dictionary of arrays.
+            The type depends on the type of each example in the batch.
+        """
+        if len(batch) == 0:
+            raise ValueError('batch is empty')
+
+        first_elem = batch[0]
+
+        if device is not None and device >= 0 and self.stream is None:
+            with cuda.get_device_from_id(device):
+                self.stream = cuda.Stream(non_blocking=True)
+        self.device = device
+
+        if isinstance(first_elem, tuple):
+            result = []
+            if not isinstance(padding, tuple):
+                padding = [padding] * len(first_elem)
+
+            for i in six.moves.range(len(first_elem)):
+                result.append(self.conveyor(i)(_concat_arrays(
+                    [example[i] for example in batch], padding[i])))
+
+            for i in six.moves.range(len(first_elem)):
+                self.conveyor(i).synchronize()
+
+            return tuple(result)
+
+        elif isinstance(first_elem, dict):
+            result = {}
+            if not isinstance(padding, dict):
+                padding = {key: padding for key in first_elem}
+
+            for key in first_elem:
+                result[key] = self.conveyor(key)(_concat_arrays(
+                    [example[key] for example in batch], padding[key]))
+
+            for key in first_elem:
+                self.conveyor(key).synchronize()
+
+            return result
+
+        else:
+            return to_device(device, _concat_arrays(batch, padding))
+
+    def conveyor(self, key):
+        if key not in self._conveyor:
+            self._conveyor[key] = Conveyor(self.device, self.stream)
+        return self._conveyor[key]
+
+
+class Conveyor(object):
+
+    """Interface to handle asynchronous data transfer using double buffering.
+
+    Args:
+        device (int): Device ID to which an array is sent. Negative value
+            indicates the host memory (CPU). If it is omitted, the array is
+            left in the original device. Asynchronous data transfer is used
+            only when device ID >= 0.
+        stream (cupy.cuda.Stream): CUDA stream. An array is sent to GPU
+            asynchronously using this stream. If ``None``, asynchronous data
+            transfer is not used.
+    """
+
+    def __init__(self, device=None, stream=None):
+        self.device = device
+        self.stream = stream
+        self.array_set = [[None, None], [None, None]]
+
+    def __call__(self, array):
```

Comment:
I feel the following API is easier to understand.
- Rename `__call__` to `put`, which does not return anything and just launch asynchronous host-to-device copy (followed by allocations if needed).
- Rename `synchronize` to `get` and let it return the target cupy array with synchronization.

Reply:
I agree with you. changed API in that way.